src/azure/cli/commands/arm.py

```diff
@@ -150,3 +151,201 @@ def required_values_validator(namespace):
+
+index_regex = re.compile(r'\[(.*)\]')
+def set_properties(instance, expression):
+    key, value = expression.split('=', 1)
+
+    #pylint:disable=redefined-variable-type
+    if value in ('{}', '{ }'):
+        value = {}
+    elif value in ('[]', '[ ]'):
+        value = []
+    elif value.lower() == 'false':
+        value = False
+    elif value.lower() == 'true':
+        value = True
+
+    name, path = _get_name_path(key)
+    instance = _find_property(instance, path)
+    match = index_regex.match(name)
+    index_value = int(match.group(1)) if match else None
+    try:
+        if index_value is not None:
+            instance[index_value] = value
+        elif isinstance(instance, dict):
+            instance[name] = value
+        else:
+            setattr(instance, name, value)
+    except IndexError:
+        raise CLIError('index {} doesn\'t exist on {}'.format(index_value, _make_camel_case(name)))
+    except (AttributeError, KeyError):
+        show_options(instance, name, key.split('.'))
+
+def add_properties(instance, argument_values):
+    # The first argument indicates the path to the collection to add to.
+    list_attribute_path = _get_internal_path(argument_values.pop(0))
+    list_to_add_to = _find_property(instance, list_attribute_path)
+
+    new_value = defaultdict(lambda: {})
+    for expression in argument_values:
+        set_properties(new_value, expression)
+    list_to_add_to.append(new_value)
+
+def remove_properties(instance, argument_values):
+    # The first argument indicates the path to the collection to add to.
+    argument_values = argument_values if isinstance(argument_values, list) else [argument_values]
+
+    list_attribute_path = _get_internal_path(argument_values.pop(0))
+    list_index = None
+    try:
+        list_index = argument_values.pop(0)
+    except IndexError:
+        pass
+
+    if not list_index:
+        _find_property(instance, list_attribute_path)
+        parent_to_remove_from = _find_property(instance, list_attribute_path[:-1])
+        del parent_to_remove_from[list_attribute_path[-1]]
+    else:
+        list_to_remove_from = _find_property(instance, list_attribute_path)
+        try:
+            list_to_remove_from.pop(int(list_index))
+        except IndexError:
+            raise CLIError('index {} doesn\'t exist on {}'
+                           .format(list_index,
+                                   _make_camel_case(list_attribute_path[-1])))
+
+def show_options(instance, part, path):
+    options = instance.__dict__ if hasattr(instance, '__dict__') else instance
+    options = options.keys() if isinstance(options, dict) else options
+    options = [_make_camel_case(x) for x in options]
+    raise CLIError('Couldn\'t find "{}" in "{}".  Available options: {}'
+                   .format(_make_camel_case(part),
+                           _make_camel_case('.'.join(path[:-1]).replace('.[', '[')),
+                           sorted(list(options), key=str)))
```

Comment:
I like the idea of having better support for objects that aren't populated, but I think specifying a tree of objects to reflect on instead of the objects themselves can be added as an improvement instead of in this PR.  I'll file an issue for it.

---
In reply to: [72831389](https://github.com/Azure/azure-cli/pull/558#discussion_r72831389) [](ancestors = 72831389)